include/matrix_assembler.h

```diff
@@ -23,89 +18,180 @@ class MatrixAssembler {
 
  public:
   //! Constructor
-  explicit MatrixAssembler(bool pdd_mode = false);
+  explicit MatrixAssembler(const std::shared_ptr<Mesh>& mesh,
+                           bool pdd_mode = false)
+      : mesh_{mesh}, pdd_{pdd_mode} {
+    nnodes_ = mesh_->nnodes();
+    nlinks_ = mesh_->nlinks();
+    init_variable_vector();
+    assemble_balance_headloss_matrix();
+    initialize_jacobian();
+  }
 
   //! Destructor
   ~MatrixAssembler() = default;
 
-  //! Obtain global nodal and pipe indices and pointers from meshes
-  //! \param[in] mesh meshes that form the pipe network
-  void global_nodal_pipe_indices(const std::shared_ptr<Mesh>& mesh);
-
-  //! Return number of nodes in the network
-  //! \retval nnode_ number of nodes in the network
-  unsigned nnodes() { return nnode_; }
-
-  //! Return number of pipes in the network
-  //! \retval nnode_ number of pipes in the network
-  unsigned npipes() { return npipe_; }
-
-  //! Initialize variable vector
-  void assemble_variable_vector();
-
-  //! Assemble Jacobian matrix for nodal head, demand and pipe discharge as
-  //! variables
-  void assemble_jacobian();
-
-  //! Apply variables (head, demand and discharge) to nodes and pipes
-  void apply_variables();
-
-  //! Initialize variable vector (unknown variables only)
-  void assemble_residual_vector();
-
-  //! Return variable vector
-  //! \retval variable_vec_ pointer to variable vector
-  std::shared_ptr<Eigen::VectorXd> variable_vec() const {
+  //! get the variable vector
+  std::shared_ptr<Eigen::VectorXd> variable_vector() const {
```

Comment:
so this is not going tobe modified?

Reply:
The pointer will be copied inside the solver class. In other word, solver class will have its own pointer to variable and residual